sparse/file.go

```diff
@@ -1,227 +1,244 @@
 package sparse
 
 import (
-	"crypto/sha1"
+	"crypto/sha512"
+	"fmt"
+	"io"
 	"os"
+	"syscall"
+	"unsafe"
 
-	"fmt"
+	log "github.com/Sirupsen/logrus"
+)
 
-	"sync"
+type FileIoProcessor interface {
+	// File I/O methods for direct or bufferend I/O
+	fileReadAt(data []byte, offset int64) (int, error)
+	fileWriteAt(data []byte, offset int64) (int, error)
+	getFile() *os.File
+	fileClose()
+	fileSync()
+	fileTruncate(size int64) error
+	fileSeek(offset int64, whence int) (ret int64, err error)
+	fileName() string
+}
 
-	fio "github.com/rancher/sparse-tools/directfio"
-	"github.com/rancher/sparse-tools/log"
-)
+type BufferedFileIoProcessor struct {
+	*os.File
+}
+
+func NewBufferedFileIoProcessor(name string, flag int, perm os.FileMode, isCreate ...bool) (*BufferedFileIoProcessor, error) {
+	file, err := os.OpenFile(name, flag, perm)
 
-// File I/O methods for direct or bufferend I/O
-var fileOpen func(name string, flag int, perm os.FileMode) (*os.File, error)
-var fileReadAt func(file *os.File, data []byte, offset int64) (int, error)
-var fileWriteAt func(file *os.File, data []byte, offset int64) (int, error)
-
-func fileBufferedOpen(name string, flag int, perm os.FileMode) (*os.File, error) {
-	return os.OpenFile(name, flag, perm)
-}
-func fileBufferedReadAt(file *os.File, data []byte, offset int64) (int, error) {
-	return file.ReadAt(data, offset)
-}
-func fileBufferedWriteAt(file *os.File, data []byte, offset int64) (int, error) {
-	return file.WriteAt(data, offset)
-}
-
-// SetupFileIO Sets up direct file I/O or buffered for small unaligned files
-func SetupFileIO(direct bool) {
-	if direct {
-		fileOpen = fio.OpenFile
-		fileReadAt = fio.ReadAt
-		fileWriteAt = fio.WriteAt
-		log.Info("Mode: directfio")
-	} else {
-		fileOpen = fileBufferedOpen
-		fileReadAt = fileBufferedReadAt
-		fileWriteAt = fileBufferedWriteAt
-		log.Info("Mode: buffered")
+	// if file does not exist, we need to create it if asked to
+	if err != nil && len(isCreate) > 0 && isCreate[0] {
+		file, err = os.Create(name)
+	}
+	if err != nil {
+		return nil, err
 	}
+
+	return &BufferedFileIoProcessor{file}, nil
+}
+
+func NewBufferedFileIoProcessorByFP(fp *os.File) *BufferedFileIoProcessor {
+	return &BufferedFileIoProcessor{fp}
+}
+
+func (file *BufferedFileIoProcessor) fileReadAt(data []byte, offset int64) (int, error) {
+	return file.File.ReadAt(data, offset)
+}
+
+func (file *BufferedFileIoProcessor) fileWriteAt(data []byte, offset int64) (int, error) {
+	return file.File.WriteAt(data, offset)
 }
 
-func loadFileLayout(abortStream <-chan error, file *os.File, layoutStream chan<- FileInterval, errStream chan<- error) error {
-	size, err := file.Seek(0, os.SEEK_END)
+func (file *BufferedFileIoProcessor) getFile() *os.File {
+	return file.File
+}
+
+func (file *BufferedFileIoProcessor) fileClose() {
```

Comment:
In fact you don't need these boiler-plate function if you keep the interface of FileIoProcessor aligned with os.File. Because os.File already implemented those and they will be used for FileIoProcessor.

LIke this: 
```
package main

import (
	"fmt"
	"os"
)

type FileIOProcessor interface {
	// File I/O methods for direct or bufferend I/O
	fileReadAt(data []byte, offset int64) (int, error)
	fileWriteAt(data []byte, offset int64) (int, error)
	getFile() *os.File
	Close() error
	Sync() error
	Truncate(size int64) error
	Seek(offset int64, whence int) (ret int64, err error)
	Name() string
}

type FileIO struct {
	*os.File
}

func CallFileFunction(f FileIOProcessor) string {
	return f.Name()
}

func (f *FileIO) fileReadAt(data []byte, offset int64) (int, error) {
	return 0, nil
}

func (f *FileIO) fileWriteAt(data []byte, offset int64) (int, error) {
	return 0, nil
}

func (f *FileIO) getFile() *os.File {
	return f.File
}

func main() {
	f, err := os.Open("test")
	if err != nil {
		fmt.Println(err)
	}
	fio := &FileIO{f}
	fmt.Println(CallFileFunction(fio))
}
```

Remember to distinguish ReadAt/WriteAt though.